File: ats_scanner/extractor.py
"""
Text extraction utilities.
Supports: .txt files, .pdf files (basic), .tex files (LaTeX), .docx files (Word), raw text strings.
"""

import os
import re
import struct
import zipfile
import zlib
from xml.etree import ElementTree


def extract_text(source: str) -> str:
    """
    Extract text from a file path or raw text string.
    Supports .txt, .pdf, .tex, and .docx files, or plain text input.
    """
    if os.path.isfile(source):
        ext = os.path.splitext(source)[1].lower()
        if ext == ".pdf":
            return _extract_pdf(source)
        elif ext == ".txt":
            return _extract_txt(source)
        elif ext == ".tex":
            return _extract_tex(source)
        elif ext == ".docx":
            return _extract_docx(source)
        else:
            raise ValueError(
                f"Unsupported file type: {ext}. Use .txt, .pdf, .tex, or .docx"
            )
    else:
        # Treat as raw text
        return source.strip()


def _extract_txt(path: str) -> str:
    """Read plain text file."""
    encodings = ["utf-8", "utf-8-sig", "latin-1", "cp1252"]
    for enc in encodings:
        try:
            with open(path, "r", encoding=enc) as f:
                return f.read().strip()
        except (UnicodeDecodeError, LookupError):
            continue
    raise ValueError(f"Could not decode file: {path}")


def _extract_pdf(path: str) -> str:
    """
    Basic PDF text extraction without external libraries.
    Works on most text-based PDFs. For scanned PDFs, use a .txt copy instead.
    """
    try:
        with open(path, "rb") as f:
            data = f.read()

        text_parts = []

        # Extract text from BT...ET blocks
        bt_pattern = re.compile(rb"BT(.*?)ET", re.DOTALL)
        tj_pattern = re.compile(rb"\(([^)]*)\)\s*T[Jj]")
        tj_array = re.compile(rb"\[([^\]]*)\]\s*TJ")
        str_in_array = re.compile(rb"\(([^)]*)\)")

        for bt_match in bt_pattern.finditer(data):
            block = bt_match.group(1)

            # Direct string Tj
            for m in tj_pattern.finditer(block):
                try:
                    text_parts.append(m.group(1).decode("latin-1", errors="ignore"))
                except Exception:
                    pass

            # Array TJ
            for m in tj_array.finditer(block):
                for sm in str_in_array.finditer(m.group(1)):
                    try:
                        text_parts.append(
                            sm.group(1).decode("latin-1", errors="ignore")
                        )
                    except Exception:
                        pass

        raw = " ".join(text_parts)

        # Clean up PDF escape sequences
        raw = raw.replace("\\n", "\n").replace("\\r", "\n").replace("\\t", " ")
        raw = re.sub(r"\\([0-7]{3})", lambda m: chr(int(m.group(1), 8)), raw)
        raw = re.sub(r"\s+", " ", raw).strip()

        if len(raw) < 50:
            raise ValueError(
                "PDF text extraction yielded too little text. Try saving as .txt first."
            )

        return raw

    except Exception as e:
        raise ValueError(
            f"Could not extract text from PDF: {e}\n"
            "Tip: Copy your resume text into a .txt file for best results."
        )


def _extract_tex(path: str) -> str:
    raw = _extract_txt(path)
    return strip_latex(raw)


def strip_latex(text: str) -> str:
    # Remove comments (% to end of line, but not escaped \%)
    text = re.sub(r"(?<!\\)%.*", "", text)

    # Remove \begin{...} and \end{...}
    text = re.sub(r"\\(?:begin|end)\{[^}]*\}", "", text)

    # Remove preamble commands and their arguments
    text = re.sub(
        r"\\(?:documentclass|usepackage|input|include|bibliography|bibliographystyle)"
        r"(?:\[[^\]]*\])?\{[^}]*\}",
        "",
        text,
    )

    # Convert \item to newline (before generic command removal)
    text = re.sub(r"\\item\s*", "\n", text)

    # Multi-pass: peel nested formatting commands layer by layer
    _fmt_re = re.compile(
        r"\\(?:textbf|textit|emph|underline|texttt|textrm|textsf|textsc"
        r"|section|subsection|subsubsection|paragraph|subparagraph"
        r"|title|author|date)\*?\{([^}]*)\}"
    )
    for _ in range(3):
        text, n = _fmt_re.subn(r"\1", text)
        if n == 0:
            break

    # Handle \href{url}{text} -> text
    text = re.sub(r"\\href\{[^}]*\}\{([^}]*)\}", r"\1", text)

    # Strip moderncv / resume-class command names but KEEP their {arg} content.
    # e.g. \cvitem{Key}{Value} -> {Key}{Value} (braces cleaned later)
    #      \cventry{yr}{title}{co}{loc}{grade}{details} -> keeps all args
    #      \name{First}{Last} -> {First}{Last}
    text = re.sub(
        r"\\(?:cvitem|cventry|cvlistitem|cvdoubleitem|cvcolumn"
        r"|cvskill|cvitemwithcomment"
        r"|name|address|email|phone|social|extrainfo)"
        r"(?:\[[^\]]*\])?",
        " ",
        text,
    )

    # Remove remaining \command with optional [] and {} args
    text = re.sub(r"\\[a-zA-Z]+\*?(?:\[[^\]]*\])?(?:\{[^}]*\})*", "", text)

    # Replace escaped special characters with space
    text = re.sub(r"\\[\\&%$#_{}~^]", " ", text)

    # Clean up leftover braces, multiple spaces, blank lines
    text = re.sub(r"[{}]", " ", text)
    text = re.sub(r"[ \t]+", " ", text)
    text = re.sub(r"\n\s*\n+", "\n", text)

    return text.strip()


def _extract_docx(path: str) -> str:
    """
    Extract text from a .docx file without external libraries.
    .docx is a ZIP archive; text content is in word/document.xml.
    """
    try:
        with zipfile.ZipFile(path, "r") as z:
            if "word/document.xml" not in z.namelist():
                raise ValueError("Invalid .docx file: word/document.xml not found")
            xml_content = z.read("word/document.xml")

        tree = ElementTree.fromstring(xml_content)

        # Word XML namespace
        W_NS = "http://schemas.openxmlformats.org/wordprocessingml/2006/main"

        paragraphs = []
        for para in tree.iter(f"{{{W_NS}}}p"):
            texts = []
            for node in para.iter(f"{{{W_NS}}}t"):
                if node.text:
                    texts.append(node.text)
            if texts:
                paragraphs.append("".join(texts))

        result = "\n".join(paragraphs).strip()

        if len(result) < 10:
            raise ValueError(
                "DOCX text extraction yielded too little text. "
                "Try saving as .txt first."
            )

        return result

    except zipfile.BadZipFile:
        raise ValueError(
            f"Could not read .docx file: {path}\n"
            "The file may be corrupted. Try re-saving from Word or use .txt."
        )
    except ValueError:
        raise
    except Exception as e:
        raise ValueError(
            f"Could not extract text from DOCX: {e}\n"
            "Tip: Copy your resume text into a .txt file for best results."
        )


def clean_text(text: str) -> str:
    text = text.lower()
    text = re.sub(r"[^\w\s\+#\./]", " ", text)
    text = re.sub(r"\s+", " ", text)
    return text.strip()

File: apps/cli/src/extractor.ts
import { readFileSync } from "node:fs";
import { extname } from "node:path";
import { createInflateRaw } from "node:zlib";
import { stripLatex } from "@ats-scanner/core";

export function extractText(source: string): string {
  const ext = extname(source).toLowerCase();
  switch (ext) {
    case ".pdf":
      return extractPdf(source);
    case ".txt":
      return extractTxt(source);
    case ".tex":
      return extractTex(source);
    case ".docx":
      return extractDocx(source);
    default:
      throw new Error(
        `Unsupported file type: ${ext}. Use .txt, .pdf, .tex, or .docx`,
      );
  }
}

function extractTxt(path: string): string {
  const encodings: BufferEncoding[] = ["utf-8", "latin1"];
  for (const enc of encodings) {
    try {
      return readFileSync(path, { encoding: enc }).trim();
    } catch {
      continue;
    }
  }
  throw new Error(`Could not decode file: ${path}`);
}

function extractPdf(path: string): string {
  const data = readFileSync(path);
  const raw = data.toString("latin1");
  const parts: string[] = [];

  const btPattern = /BT([\s\S]*?)ET/g;
  const tjPattern = /\(([^)]*)\)\s*T[Jj]/g;
  const tjArray = /\[([^\]]*)\]\s*TJ/g;
  const strInArray = /\(([^)]*)\)/g;

  let btMatch: RegExpExecArray | null;
  while ((btMatch = btPattern.exec(raw)) !== null) {
    const block = btMatch[1];

    let tjMatch: RegExpExecArray | null;
    const tjPat = new RegExp(tjPattern.source, tjPattern.flags);
    while ((tjMatch = tjPat.exec(block)) !== null) {
      parts.push(tjMatch[1]);
    }

    let arrMatch: RegExpExecArray | null;
    const arrPat = new RegExp(tjArray.source, tjArray.flags);
    while ((arrMatch = arrPat.exec(block)) !== null) {
      let sm: RegExpExecArray | null;
      const sPat = new RegExp(strInArray.source, strInArray.flags);
      while ((sm = sPat.exec(arrMatch[1])) !== null) {
        parts.push(sm[1]);
      }
    }
  }

  let text = parts.join(" ");
  text = text.replace(/\\n/g, "\n").replace(/\\r/g, "\n").replace(/\\t/g, " ");
  text = text.replace(/\\([0-7]{3})/g, (_, oct) =>
    String.fromCharCode(parseInt(oct, 8)),
  );
  text = text.replace(/\s+/g, " ").trim();

  if (text.length < 50) {
    throw new Error(
      "PDF text extraction yielded too little text. Try saving as .txt first.",
    );
  }

  return text;
}

function extractTex(path: string): string {
  const raw = extractTxt(path);
  return stripLatex(raw);
}

function extractDocx(path: string): string {
  const data = readFileSync(path);
  const bytes = new Uint8Array(data.buffer, data.byteOffset, data.byteLength);

  let eocdOffset = -1;
  for (
    let i = bytes.length - 22;
    i >= Math.max(0, bytes.length - 65557);
    i--
  ) {
    if (
      bytes[i] === 0x50 &&
      bytes[i + 1] === 0x4b &&
      bytes[i + 2] === 0x05 &&
      bytes[i + 3] === 0x06
    ) {
      eocdOffset = i;
      break;
    }
  }
  if (eocdOffset < 0) throw new Error("Not a valid ZIP/DOCX file");

  const view = new DataView(
    data.buffer,
    data.byteOffset,
    data.byteLength,
  );
  const cdOffset = view.getUint32(eocdOffset + 16, true);
  const cdSize = view.getUint32(eocdOffset + 12, true);

  let pos = cdOffset;
  let docEntry: {
    compMethod: number;
    compSize: number;
    localOffset: number;
  } | null = null;

  while (pos < cdOffset + cdSize) {
    if (view.getUint32(pos, true) !== 0x02014b50) break;
    const compMethod = view.getUint16(pos + 10, true);
    const compSize = view.getUint32(pos + 20, true);
    const nameLen = view.getUint16(pos + 28, true);
    const extraLen = view.getUint16(pos + 30, true);
    const commentLen = view.getUint16(pos + 32, true);
    const localOffset = view.getUint32(pos + 42, true);
    const name = new TextDecoder().decode(bytes.slice(pos + 46, pos + 46 + nameLen));

    if (name === "word/document.xml") {
      docEntry = { compMethod, compSize, localOffset };
      break;
    }
    pos += 46 + nameLen + extraLen + commentLen;
  }

  if (!docEntry) throw new Error("word/document.xml not found in .docx");

  const localNameLen = view.getUint16(docEntry.localOffset + 26, true);
  const localExtraLen = view.getUint16(docEntry.localOffset + 28, true);
  const dataStart = docEntry.localOffset + 30 + localNameLen + localExtraLen;
  const compData = bytes.slice(dataStart, dataStart + docEntry.compSize);

  let xmlText: string;
  if (docEntry.compMethod === 0) {
    xmlText = new TextDecoder().decode(compData);
  } else {
    const inflated = inflateRawSync(Buffer.from(compData));
    xmlText = new TextDecoder().decode(inflated);
  }

  const ns = "http://schemas.openxmlformats.org/wordprocessingml/2006/main";
  const paragraphs: string[] = [];

  const pRegex = new RegExp(
    `<[^>]*?:?p(?:\\s[^>]*)?>([\\s\\S]*?)<\\/[^>]*?:?p>`,
    "g",
  );
  const tRegex = new RegExp(
    `<[^>]*?:?t(?:\\s[^>]*)?>([\\s\\S]*?)<\\/[^>]*?:?t>`,
    "g",
  );

  let pMatch: RegExpExecArray | null;
  while ((pMatch = pRegex.exec(xmlText)) !== null) {
    const pContent = pMatch[1];
    const texts: string[] = [];
    let tMatch: RegExpExecArray | null;
    const tPat = new RegExp(tRegex.source, tRegex.flags);
    while ((tMatch = tPat.exec(pContent)) !== null) {
      texts.push(tMatch[1]);
    }
    if (texts.length) paragraphs.push(texts.join(""));
  }

  const result = paragraphs.join("\n").trim();
  if (result.length < 10) {
    throw new Error(
      "DOCX text extraction yielded too little text. Try saving as .txt first.",
    );
  }

  return result;
}

function inflateRawSync(data: Buffer): Buffer {
  const { inflateRawSync: inflate } = require("node:zlib") as typeof import("node:zlib");
  return inflate(data);
}
